### PR TITLE
Refactors templates and adds QC variable attributes

### PR DESCRIPTION
**Pull request by benjwadams**
Refactors template by moving template boilerplate into a loop.  Also
adds default attributes for a number of QC variables.  Requires updated
logic from glider-dac repo.

### CODE REVIEW

**Comment by benjwadams** [comment]
Required by https://github.com/ioos/glider-dac/pull/105

File: erddap/templates/private/dataset.deployment.xml [hunk 1]
```diff
@@ -301,184 +301,35 @@
...
+            {% endif %}
         </addAttributes>
     </dataVariable>
+    {% endfor %}
 
-    {% if include_qartod %}
-    {% for var in qartod_variables %}
+    {# if there are QARTOD variables defined, populate them #}
+    {% if qc_var_types['qartod'] | length > 0 %}
+    {% for qartod_var in (qc_var_types['qartod']) %}
```
Comment (benjwadams) [inline erddap/templates/private/dataset.deployment.xml]: Dunno why I put parentheses here.

File: erddap/templates/private/dataset.deployment.xml [hunk 2]
```diff
@@ -301,184 +301,35 @@
         </addAttributes>
     </dataVariable>
 
+    {% for req_var in reqd_qc_vars %}
     <dataVariable>
-        <sourceName>time_qc</sourceName>
-        <destinationName>precise_time_qc</destinationName>
+        <sourceName>{{req_var}}</sourceName>
+        <destinationName>{{dest_var_remaps.get(req_var, req_var)}}</destinationName>
```
Comment (benjwadams) [inline erddap/templates/private/dataset.deployment.xml]: Perhaps I should have added a comment here.  This tries to get any name remappings and defaults to the original variable name if not found.  defaultdict didn't seem to cover this use case elegantly, but maybe I was missing something.

File: erddap/templates/private/dataset.deployment.xml [hunk 3]
```diff
@@ -301,184 +301,35 @@
...
-    <dataVariable>
-        <sourceName>v_qc</sourceName>
-        <destinationName>v_qc</destinationName>
-        <dataType>byte</dataType>
-        <addAttributes>
-            <att name="ioos_category">Quality</att>
-            <att name="long_name">v Variable Quality Flag</att>
+            <att name="long_name">{{dest_var_remaps.get(req_var, req_var)}} Variable Quality Flag</att>
+            {% if '_FillValue' not in qc_var_types['gen_qc'][req_var] %}
+            <att name='_FillValue'>-127</att>
```
Comment (Bobfrat) [inline erddap/templates/private/dataset.deployment.xml]: Do you need to set a type here?
```
<att name=“_FillValue" type="int">-127</att>
```
https://coastwatch.pfeg.noaa.gov/erddap/download/setupDatasetsXml.html#addAttributes

File: erddap/templates/private/dataset.deployment.xml [hunk 4]
```diff
@@ -301,184 +301,35 @@
...
-    </dataVariable>
-
-    <dataVariable>
-        <sourceName>v_qc</sourceName>
-        <destinationName>v_qc</destinationName>
-        <dataType>byte</dataType>
-        <addAttributes>
-            <att name="ioos_category">Quality</att>
-            <att name="long_name">v Variable Quality Flag</att>
+            <att name="long_name">{{dest_var_remaps.get(req_var, req_var)}} Variable Quality Flag</att>
```
Comment (Bobfrat) [inline erddap/templates/private/dataset.deployment.xml]: Does the output from ```{{dest_var_remaps.get(req_var, req_var)}}``` result in the variable name + '_qc'? If so I think we need to drop the ```_qc``` in the long name.

E.G.
```
<destinationName>u_qc</destinationName>
<att name="long_name">u Variable Quality Flag</att>

```

File: erddap/templates/private/dataset.deployment.xml [hunk 5]
```diff
@@ -301,184 +301,35 @@
...
-    </dataVariable>
-
-    <dataVariable>
-        <sourceName>v_qc</sourceName>
-        <destinationName>v_qc</destinationName>
-        <dataType>byte</dataType>
-        <addAttributes>
-            <att name="ioos_category">Quality</att>
-            <att name="long_name">v Variable Quality Flag</att>
+            <att name="long_name">{{dest_var_remaps.get(req_var, req_var).replace('_qc', '')}} Variable Quality Flag</att>
```
Comment (benjwadams) [inline erddap/templates/private/dataset.deployment.xml]: Ew.  May refactor later; too much logic in the templates.

**Comment by kwilcox** [comment]
I would recommend adding both `_FillValue` and `missing_value` attributes to maintain backwards and forward compatibility. 

I'm also not a fan of assigning the QARTOD flag of 2 for the `missing_value`. Accessing the data with (most) client libraries is going to mask the data and it will be an extra step before analysis to retrieve the actual flag value of 2.. @rsignell-usgs @ocepaf @shane-axiom any thoughts?

File: erddap/templates/private/dataset.deployment.xml [hunk 6]
```diff
@@ -301,184 +301,35 @@
...
+        <sourceName>{{qartod_var}}</sourceName>
+        <destinationName>{{qartod_var}}</destinationName>
         <dataType>byte</dataType>
         <addAttributes>
             <att name="ioos_category">Quality</att>
+            {# all qartod variables should have _FillValue, missing_value,
+               flag_values, and flag_meanings defined #}
+           <att name="flag_values">1B, 2B, 3B, 4B, 9B</att>
+           <att name="flag_meanings">GOOD NOT_EVALUATED SUSPECT BAD MISSING</att>
+           <att name="_FillValue" type='byte'>2</att>
```
Comment (Bobfrat) [inline erddap/templates/private/dataset.deployment.xml]: @kerfoot what's your opinion on setting the _FillValue to a 2 (NOT_EVALUATED) flag?

**Comment by kerfoot** [comment]
I agree with @kwilcox. The _FillValue should be a value not listed in variable:flag_values

File: erddap/templates/private/dataset.deployment.xml [hunk 7]
```diff
@@ -309,16 +309,20 @@
         <addAttributes>
             <att name="ioos_category">Quality</att>
             <att name="long_name">{{dest_var_remaps.get(req_var, req_var).replace('_qc', '')}} Variable Quality Flag</att>
-            {% if '_FillValue' not in qc_var_types['gen_qc'][req_var] %}
-            <att name='_FillValue' type='byte'>-127</att>
+            {% if req_var not in qc_var_types['gen_qc'] %}
+            <att name="flag_values">0,1,2,3,4,5,6,7,8,9</att>
```
Comment (kwilcox) [inline erddap/templates/private/dataset.deployment.xml]: Why do the `gen_qc` variables have a different syntax for `flag_values` when compared to the `qartod` variables?

`<att name="flag_values">0,1,2,3,4,5,6,7,8,9</att>` 
vs
`<att name="flag_values">1B, 2B, 3B, 4B, 9B</att>`

Comment (benjwadams) [inline erddap/templates/private/dataset.deployment.xml]: Fixed in latest to comply with CF.

File: erddap/templates/private/dataset.deployment.xml [hunk 8]
```diff
@@ -310,7 +310,7 @@
             <att name="ioos_category">Quality</att>
             <att name="long_name">{{dest_var_remaps.get(req_var, req_var).replace('_qc', '')}} Variable Quality Flag</att>
             {% if req_var not in qc_var_types['gen_qc'] %}
-            <att name="flag_values">0,1,2,3,4,5,6,7,8,9</att>
+            <att name="flag_values">0b, 1b, 2b, 3b, 4b, 5b, 6b, 7b, 8b, 9b</att>
```
Comment (kwilcox) [inline erddap/templates/private/dataset.deployment.xml]: I'm mobile and can't test this in ERDDAP or the NCML syntax at the moment... but I believe this should be:

`<att name="flag_values" type="byte">0 1 2 3 4 5 6 7 8 9</att>`

I don't know where the `b` is coming from and spaces are used as the separator in NcML/NetCDF array attributes.

Can you add any tests for these templates? Maybe loading one into NetCDF Java with some fill data and testing the attributes?

File: erddap/templates/private/dataset.deployment.xml [hunk 9]
```diff
@@ -331,7 +331,7 @@
             <att name="ioos_category">Quality</att>
             {# all qartod variables should have _FillValue, missing_value,
                flag_values, and flag_meanings defined #}
-           <att name="flag_values">1B, 2B, 3B, 4B, 9B</att>
+           <att name="flag_values">1b, 2b, 3b, 4b, 9b</att>
```
Comment (kwilcox) [inline erddap/templates/private/dataset.deployment.xml]: Same as other comment

File: erddap/templates/private/dataset.deployment.xml (hunk 8 above)
Comment (Bobfrat) [inline erddap/templates/private/dataset.deployment.xml]: I agree, but I would think it needs to be a byteList, right?
```
<att name="flag_values" type="byteList">0 1 2 3 4 5 6 7 8 9</att>
```

Comment (kwilcox) [inline erddap/templates/private/dataset.deployment.xml]: It should be `byte`. `byteList` isn't a thing in the NcML Schema.  See: http://www.unidata.ucar.edu/software/thredds/current/netcdf-java/ncml/AnnotatedSchema4.html, the "attribute Element" section and the "DataType Type" attribute.

Comment (Bobfrat) [inline erddap/templates/private/dataset.deployment.xml]: Oh I thought this was an ERDDAP datasets.xml template so I was looking at the following:
https://coastwatch.pfeg.noaa.gov/erddap/download/setupDatasetsXml.html#addAttributes

Comment (kwilcox) [inline erddap/templates/private/dataset.deployment.xml]: My mistake, you are right, I thought this was an NcML file! `byteList` is correct.

**Pull request merged**